feathr_project/feathr/registry/_feathr_registry_client.py

```diff
@@ -0,0 +1,631 @@
+            from_context: whether the feature is from context (i.e. end users has to callFeathrClient.build_features()) or the feature is from a pre-built config file. Currently Feathr only supports register features from context.
+            anchor_list: The anchor list after feature build
+            derived_feature_list: the derived feature list after feature build
+        """
+        if not from_context:
+            raise RuntimeError("Currently Feathr only supports registering features from context (i.e. you must call FeathrClient.build_features() before calling this function).")
+        
+        # Before starting, create the project
+        self.project_id = self._create_project()
+
+        for anchor in anchor_list:
+            source = anchor.source
+            # 1. Create Source on the registry
+            if not hasattr(source, "_registry_id"):
+                source._registry_id = self._create_source(source)
+            # 2. Create Anchor on the registry
+            if not hasattr(anchor, "_registry_id"):
+                anchor._registry_id = self._create_anchor(anchor)
+            # 3. Create all features on the registry
+            for feature in anchor.features:
+                if not hasattr(feature, "_registry_id"):
+                    feature._registry_id = self._create_anchor_feature(
+                        anchor._registry_id, feature)
+        # 4. Create all derived features on the registry
+        for df in derived_feature_list:
+            if not hasattr(df, "_registry_id"):
+                df._registry_id = self._create_derived_feature(df)
+
+    def list_registered_features(self, project_name: str) -> List[str]:
+        """List all the already registered features. If project_name is not provided or is None, it will return all
+        the registered features; otherwise it will only return features under this project
+        """
+        resp = self._get(f"/projects/{project_name}/features")
+        # In V1 API resp should be an array, will be changed in V2 API
+        return [r["attributes"]["qualifiedName"] for r in resp]
+
+    def get_features_from_registry(self, project_name: str) -> Tuple[List[FeatureAnchor], List[DerivedFeature]]:
+        """
+        [Sync Features from registry to local workspace, given a project_name, will write project's features from registry to to user's local workspace]
+
+        Args:
+            project_name (str): project name.
+        """
+        lineage = self._get(f"/projects/{project_name}")
+        return dict_to_project(lineage)
+
+    def _create_project(self) -> UUID:
+        r = self._post(f"/projects", {"name": self.project_name})
+        self.project_id = UUID(r["guid"])
+        return self.project_id
+
+    def _create_source(self, s: Source) -> UUID:
+        r = self._post(
+            f"/projects/{self.project_id}/datasources", source_to_def(s))
+        id = UUID(r["guid"])
+        s._registry_id = id
```

Comment:
Would a `Source` class have a `_registry_id` attribute? I don't see it in the `Source` definition?

Reply:
All these `_***_id` properties are temporary during the `register_features` process.

Entities have relations, and the relations are represented as connections between GUIDs on the registry server side, so I need to record registry_id for each entity to build correct connections between them, otherwise I have to retrieve GUID from the server side by their qualified names one by one, which could be very slow in some cases.

So all these `_***_id` are just for performance boosting solely, they have no use outside of this scope so I don't think we need to add them into the public interface.